core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -230,72 +136,183 @@ protected TwinUpdate mapTwinResponseToProto(byte[] responseBytes) {
-            final T value = objectMapper.readValue(objValue, clazz);
-            LOG.trace("Updated consumer with key {}", key);
-            consumer.accept(value);
+        /**
+         * Consume the subscription.
+         *
+         * Adds a consumer to the subscription. Incoming values will be forwarded to the passed consumer until the
+         * returned value is closed. If there is a value already available for the subscription, the consumer will be
+         * called with this value immediately.
+         *
+         * @param clazz The class of the value to consume
+         * @param consumer The consumer accepting the values
+         * @param <T> The class of the value to consume
+         * @return a Closable, stopping the consumption when closed
+         */
+        public synchronized <T> Closeable consume(final Class<T> clazz, final Consumer<T> consumer) {
+            final Consumer<JsonNode> jsonConsumer = (json) -> {
+                try {
+                    // Deserialize to the final class
+                    final var value = AbstractTwinSubscriber.this.objectMapper.treeToValue(json, clazz);
+
+                    // Forward to typed consumer
+                    consumer.accept(value);
+
+                } catch (final JsonProcessingException e) {
+                    LOG.error("Deserialize twin failed: {} as {}", this.key, clazz, e);
+                }
+            };
+
+            if (this.value == null) {
+                // Initially request value
+                this.request();
+            } else {
+                // If value already exists, forward to consumer without requesting
+                jsonConsumer.accept(this.value.value);
+            }
+
+            // Add the consumer to the subscription
+            this.consumers.add(jsonConsumer);
+
+            // Return the closable removing the consumer
+            return () -> {
+                this.consumers.remove(jsonConsumer);
+
+                // Drop the subscription if it becomes empty
+                if (this.consumers.isEmpty()) {
+                    AbstractTwinSubscriber.this.subscriptions.remove(this.key);
+                }
+            };
         }
 
-        @Override
-        public boolean equals(Object o) {
-            if (this == o) return true;
-            if (o == null || getClass() != o.getClass()) return false;
-            SessionImpl<?> session = (SessionImpl<?>) o;
-            return Objects.equals(key, session.key) && Objects.equals(consumer, session.consumer) && Objects.equals(clazz, session.clazz);
+        private synchronized void accept(final Value value) {
+            Objects.requireNonNull(value);
+
+            // Skip if value has not changed
+            if (this.value != null && Objects.equals(this.value.value, value.value)) {
+                return;
+            }
+
+            // Remember value
+            this.value = value;
+
+            // Cancel outstanding retry
```

Comment:
An outstanding retry should also be canceled in case the value did not change. Otherwise we will retry forever.

Reply:
Same as above.

Reply:
This is not the same as above. `void request` schedules new requests forever. An outstanding retry must be canceled (and the member set to `null`) when a value is accepted, irrespective if the value is equal to the current value or not.

Even if `void accept` cancels the current retry there is a slight chance that retries may go on forever:

- retry was scheduled
- accept is entered
- retry starts to execute and waits to enter the synchronized `request` method
- accept cancels the retry (but does not interrupt)
- accept is left
- retry enters `request`
- a request is sent and another retry is scheduled

I think we could use the interruption mechanism here. In `void request` the current thread should check if it is interrupted and in that case it would not send a request or schedule another retry.

Reply:
I've updated this to update the value container on version/session change.